text/2019-11-07-raft-witness.md

```diff
@@ -0,0 +1,85 @@
+# Raft Witness
+
+## Summary
+
+This RFC proposes a new role Witness to Raft, and describes how it could benefit us.
+
+## Motivation
+
+To tolerate 2 replicas fail, Raft needs 5 replicas, which requires a huge energy
+footprint. However, with witness Raft could only need 4 or even 3 replicas to
+tolerate 2 fail, and the price is latency becomes higher if some peers are
+isolated, or need more time to recover from 2 peers fail. Considering they are not
+common cases, we can introduce Witness into Raft.
+
+## Dtailed Design
+
+### The Original Article
+
+Witness is firstly described in this [article][reduce-the-energy-footprint], and is
+already implemented in [dragonboat](https://github.com/lni/dragonboat). According to
+the article, Witness is like normal Voter in Raft, except
+
+* Witness can't start new elections;
+* Leader only replicates metadata (term, index) to Witness without real entries.
+
+So Witness can reduce network traffic and storage usage in a Raft group.
+
+However, there are some issues about about the design. For example (described in the article),
+suppose there is a Raft group with Voters `A, B, C, D` and Witness `E`. `A, B, E` can
+hold a quorum to commit some entries. But if `A, B` fail, `C, D, E` can't elect a new
+leader forever because `E` will never start new elections, and it will always reject
+vote messages from `C, D` because it has a higher *index* or *term*. Even if we can
+make `C` or `D` become leader manually (with some external tools), entries committed by
+`A, B, E` are still lost, which is fatal for a database.
+
+### Improved
+
+Considering the issues, this RFC trys to improve the machanism to make it more suitable
+for distribution databases. The design is,
+
+* Witness itself acts totally same as normal Voters;
+* Witness votes voters if they have same index and term;
+* Generally, Leader never appends entries to Witness;
+* In some cases, Leader can append entries (NOT ONLY metadata) to Witness.
+
+The core principle is leader only try its best to avoid appending to Witness. For example,
+suppose there is a Raft group with voters `A, B` and witness `C`, and `A` is the leader.
+When A broadcasts new entries to the group, it can ignore the witness `C` if `B` is recently
+active.
+
+To show what it benefits Raft, here are 2 examples:
+
+The first is a Raft group with normal voters `A, B, C, D` and a witness `E`. It
+reduces 20% energy footprint and network traffic from the topology with 5 normal voters,
```

Comment:
It seems if you save 20% of 5 nodes it means that your reduce the network traffic by a fifth? Doesn't the leader still replicate metadata to the witness?

Reply:
In my design generally leader won't send replicate metadata to witness. However, heartbeat is kept. So It's less than 20%, I guess.

Reply:
I would expect it to be quite a bit less, since the network communication still needs to happen :(